src/ck-core/ck.C

```diff
@@ -722,22 +744,12 @@ void CkCreateLocalNodeGroup(CkGroupID groupID, int epIdx, envelope *env)
 #endif
 
   CkpvAccess(_currentNodeGroupObj) = NULL;
-  _STATS_RECORD_PROCESS_NODE_GROUP_1();
 
   CmiImmediateLock(CksvAccess(_nodeGroupTableImmLock));
-  CksvAccess(_nodeGroupTable)->find(groupID).setObj(obj);
-  CksvAccess(_nodeGroupTable)->find(groupID).setcIdx(gIdx);
-  CksvAccess(_nodeGroupIDTable).push_back(groupID);
-
-  PtrQ *ptrq = CksvAccess(_nodeGroupTable)->find(groupID).getPending();
-  if(ptrq) {
-    void *pending;
-    while((pending=ptrq->deq())!=0) {
-      _CldNodeEnqueue(CkMyNode(), pending, _infoIdx);
-    }
-    CksvAccess(_nodeGroupTable)->find(groupID).clearPending();
-  }
```

Comment:
Why separate this into two different critical sections?

Reply:
Well, it puts the object into the table so it can access itself during its creation (e.g., array reduction managers call ckLocalBranch and expect to get their associated group), but doesn't notify the pending listeners until its construction has finished (i.e., after the EP is invoked).

Reply:
Also, if the dependencies can individually specify whether or not they're node-level (which would be a trivial addition from this PR) -- both node/groups would require two critical sections.

Reply:
Ah, I see, makes sense. I think it'd be useful to have that explanation enshrined as a comment in the code.